services/spar/src/Spar/Data.hs

```diff
@@ -619,15 +623,36 @@ lookupScimToken ::
   ScimToken ->
   m (Maybe ScimTokenInfo)
 lookupScimToken token = do
-  mbRow <- retry x1 . query1 sel $ params Quorum (Identity token)
-  pure $ fmap fromScimTokenRow mbRow
-  where
-    sel :: PrepQuery R (Identity ScimToken) ScimTokenRow
+  let tokenHash = hashScimToken token
+  mbRow <- retry x1 . query1 sel $ params Quorum (tokenHash, token)
+  for mbRow $ \row -> do
+    let tokenInfo = fromScimTokenRow row
+    case scimTokenLookupKey row of
+      ScimTokenLookupKeyHashed _ -> pure ()
+      ScimTokenLookupKeyPlaintext token' ->
+        connvertPlaintextToken token' tokenInfo
+    pure tokenInfo
+  where
+    sel :: PrepQuery R (ScimTokenHash, ScimToken) ScimTokenRow
     sel =
       [r|
       SELECT token_, team, id, created_at, idp, descr
-        FROM team_provisioning_by_token WHERE token_ = ?
-    |]
+        FROM team_provisioning_by_token WHERE token_ in (?, ?)
+      |]
+
+connvertPlaintextToken ::
+  (HasCallStack, MonadClient m) =>
+  ScimToken ->
+  ScimTokenInfo ->
+  m ()
+connvertPlaintextToken token ScimTokenInfo {..} = retry x5 . batch $ do
+  setType BatchLogged
+  setConsistency Quorum
+  addPrepQuery delById (stiTeam, stiId)
+  addPrepQuery delByTokenLookup (Identity (ScimTokenLookupKeyPlaintext token))
+  let tokenHash = hashScimToken token
+  addPrepQuery insByToken (ScimTokenLookupKeyHashed tokenHash, stiTeam, stiId, stiCreatedAt, stiIdP, stiDescr)
+  addPrepQuery insByTeam (ScimTokenLookupKeyHashed tokenHash, stiTeam, stiId, stiCreatedAt, stiIdP, stiDescr)
```

Comment:
Deletion of the old entries must happen after insertion of the new ones, to handle crashes in between.  Both, or even either of the rows being hashed while there is still an unhashed
```suggestion
  addPrepQuery delByTokenLookup (Identity (ScimTokenLookupKeyPlaintext token))
  let tokenHash = hashScimToken token
  -- enter by new lookup key
  addPrepQuery insByToken (ScimTokenLookupKeyHashed tokenHash, stiTeam, stiId, stiCreatedAt, stiIdP, stiDescr)
  -- update info table
  addPrepQuery insByTeam (ScimTokenLookupKeyHashed tokenHash, stiTeam, stiId, stiCreatedAt, stiIdP, stiDescr)
  -- remove old lookup key
  addPrepQuery delByTokenLookup (Identity (ScimTokenLookupKeyPlaintext token))
```